### PR TITLE
DB authentication

### PR DESCRIPTION
**Pull request by vlykhvar**
Configure DB authentication instead of In memory authentication
Add Role entity, Dao and Service layer for it.
Use BCryptPasswordEncoder for storing passwords in secured way
Configure role access to specific resources for ADMIN and for USER. For example only ADMIN can add new movies or movie sessions.

### CODE REVIEW

File: src/main/java/com/dev/theater/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -6,23 +6,19 @@
...
 public class UserServiceImpl implements UserService {
     private final UserDao userDao;
-    private final PasswordEncoder passwordEncoder;
 
     @Autowired
-    public UserServiceImpl(UserDao userDao, PasswordEncoder passwordEncoder) {
+    public UserServiceImpl(UserDao userDao) {
         this.userDao = userDao;
-        this.passwordEncoder = passwordEncoder;
     }
```
Comment (yurwing) [inline src/main/java/com/dev/theater/service/impl/UserServiceImpl.java]: why did you remove "passwordEncoder"?

**Review by yurwing — APPROVED** [review]

File: src/main/java/com/dev/theater/confing/AppConfig.java [hunk 2]
```diff
@@ -10,14 +10,17 @@
...
 
 @Configuration
 @PropertySource("classpath:db.properties")
 @ComponentScan(basePackages = {
         "com.dev.theater.service",
         "com.dev.theater.dao",
         "com.dev.theater.security",
-        "com.dev.theater.mapper"
+        "com.dev.theater.mapper",
+        "com.dev.theater.confing"
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/AppConfig.java]:
```suggestion
        "com.dev.theater.config"
```

File: src/main/java/com/dev/theater/confing/DataInitializer.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package com.dev.theater.confing;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/DataInitializer.java]:
```suggestion
package com.dev.theater.config;
```

File: src/main/java/com/dev/theater/confing/DataInitializer.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    private final RoleService roleService;
+    private final UserService userService;
+
+    public DataInitializer(RoleService roleService, UserService userService) {
+        this.roleService = roleService;
+        this.userService = userService;
+    }
+
+    @PostConstruct
+    public void addAdmin() {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/DataInitializer.java]: let's make these 2 methods private and call them inside one `@PostConstruct` method
conceptually this is a bit confusing)

File: src/main/java/com/dev/theater/confing/SecurityConfig.java [hunk 5]
```diff
@@ -1,37 +1,56 @@
...
-                .password(passwordEncoder().encode("123")).roles("USER");
+        auth.userDetailsService(userDetailsService).passwordEncoder(passwordEncoder);
     }
 
     protected void configure(HttpSecurity http) throws Exception {
         http.authorizeRequests()
+                .antMatchers(HttpMethod.POST, "/register").permitAll()
+                .antMatchers(HttpMethod.GET, "/cinema-halls").hasRole("USER")
+                .antMatchers(HttpMethod.POST, "/cinema-halls").hasRole("ADMIN")
+                .antMatchers(HttpMethod.POST, "/movies").hasRole("USER")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/SecurityConfig.java]: only admin can create new movies

File: src/main/java/com/dev/theater/confing/SecurityConfig.java [hunk 6]
```diff
@@ -1,37 +1,56 @@
...
+                .antMatchers(HttpMethod.POST, "/movie-sessions").hasRole("ADMIN")
+                .antMatchers(HttpMethod.POST, "/movie-sessions/").hasRole("ADMIN")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/SecurityConfig.java]: are these lines different?

File: src/main/java/com/dev/theater/confing/SecurityConfig.java [hunk 7]
```diff
@@ -1,37 +1,56 @@
...
                 .and()
-                .formLogin()
-                .permitAll()
-                .and()
-                .httpBasic()
-                .and()
-                .csrf().disable();
+                .formLogin().permitAll()
+                .and().csrf().disable();
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/SecurityConfig.java]: what about .httpBasic()?

File: src/main/java/com/dev/theater/confing/SecurityConfig.java [hunk 8]
```diff
@@ -1,37 +1,56 @@
...
     }
 
     @Bean
-    public PasswordEncoder passwordEncoder() {
-        return new BCryptPasswordEncoder();
+    public UserDetailsService userDetailsService() {
+        return new UserDetailsServiceImp();
     }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/confing/SecurityConfig.java]: do you need this?

File: src/main/java/com/dev/theater/dao/impl/RoleDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    public RoleDaoImpl(SessionFactory sessionFactory) {
+        super(sessionFactory);
+    }
+
+    @Override
+    public Role getRoleByName(String roleName) {
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("from Role where roleName = :roleName", Role.class)
+                    .setParameter("roleName", roleName)
+                    .uniqueResultOptional().orElseThrow();
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/dao/impl/RoleDaoImpl.java]: can we simplify this?

File: src/main/java/com/dev/theater/dao/impl/RoleDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        } catch (Exception e) {
+            throw new CrudException("Could not find " + roleName, e);
+        }
+    }
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/dao/impl/RoleDaoImpl.java]: redundant empty line

File: src/main/java/com/dev/theater/model/User.java [hunk 11]
```diff
@@ -17,15 +19,16 @@
     private String email;
     @Column(nullable = false)
     private String password;
+    @OneToMany
     @Column(nullable = false)
-    private byte[] salt;
+    private Set<Role> roleName;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/model/User.java]:
```suggestion
    private Set<Role> roles;
```

File: src/main/java/com/dev/theater/model/User.java [hunk 12]
```diff
@@ -17,15 +19,16 @@
     private String email;
     @Column(nullable = false)
     private String password;
+    @OneToMany
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/model/User.java]: reconsider this relationship - see checklist

File: src/main/java/com/dev/theater/security/impl/AuthenticationServiceImpl.java [hunk 13]
```diff
@@ -11,19 +12,23 @@
...
         this.shoppingCartService = shoppingCartService;
+        this.roleService = roleService;
     }
 
     @Override
     public User register(String email, String password) {
         User newUser = new User();
         newUser.setEmail(email);
         newUser.setPassword(password);
+        newUser.getRoleName().add(roleService.getRoleByName("User"));
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/security/impl/AuthenticationServiceImpl.java]: do you have `User` role?

File: src/main/java/com/dev/theater/dao/impl/UserDaoImpl.java [hunk 14]
```diff
@@ -29,9 +29,11 @@ public UserDaoImpl(SessionFactory sessionFactory) {
     @Override
     public Optional<User> findByEmail(String email) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/dao/impl/UserDaoImpl.java:30]: what about findById() method? let's use join fetch there as well

File: src/main/java/com/dev/theater/security/UserDetailsServiceImp.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import com.dev.theater.service.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.User.UserBuilder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserDetailsServiceImp implements UserDetailsService {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/security/UserDetailsServiceImp.java]:
```suggestion
public class UserDetailsServiceImpl implements UserDetailsService {
```

File: src/main/java/com/dev/theater/security/UserDetailsServiceImp.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    private UserService userService;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/security/UserDetailsServiceImp.java]:
```suggestion
    private final UserService userService;
```

File: src/main/java/com/dev/theater/security/UserDetailsServiceImp.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Autowired
+    public UserDetailsServiceImp(UserService userService) {
+        this.userService = userService;
+    }
+
+    public UserDetailsServiceImp() {
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/security/UserDetailsServiceImp.java]: do you need this?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]
good job!

**Pull request merged**